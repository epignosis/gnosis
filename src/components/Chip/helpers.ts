import { MutableRefObject, useLayoutEffect, useState } from "react";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const useIsOverflowX = (overflowRef: MutableRefObject<any>): boolean => {
  const [isOverflow, setIsOverflow] = useState<boolean>(false);

  useLayoutEffect(() => {
    const el = overflowRef.current;

    const trigger = (): void => {
      setIsOverflow(el.offsetWidth < el.scrollWidth);
    };

    if (el) {
      if ("ResizeObserver" in window) {
        new ResizeObserver(trigger).observe(el);
      }

      trigger();
    }
  }, [overflowRef]);

  return isOverflow;
};
