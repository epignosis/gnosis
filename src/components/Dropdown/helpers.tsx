import { DropdownItem } from "./types";

export const filterListByKeyword = (list: DropdownItem[], keyword: string): DropdownItem[] => {
  return list.reduce((arr, item): DropdownItem[] => {
    const { items, label, originalText } = item;

    if (items && items.length > 0) {
      const filteredItems = filterListByKeyword(items, keyword);
      if (filteredItems.length > 0) {
        arr.push({ ...item, items: [...filteredItems] });
      }

      return arr;
    }

    if (typeof label === "string") {
      if (label.toLowerCase().includes(keyword.toLowerCase())) {
        arr.push(item);
      }
    } else if (originalText) {
      if (originalText.toLowerCase().includes(keyword.toLowerCase())) {
        arr.push(item);
      }
    }

    return arr;
  }, [] as DropdownItem[]);
};

// Reference: https://gist.github.com/twxia/bb20843c495a49644be6ea3804c0d775
export const getScrollableParent = (node: ParentNode | null): ParentNode | null => {
  const isElement = node instanceof HTMLElement;
  const overflowY = isElement && window.getComputedStyle(node).overflowY;
  const isScrollable = overflowY !== "visible" && overflowY !== "hidden";

  if (!node) {
    return null;
  } else if (isScrollable && isElement && node.scrollHeight >= node.clientHeight) {
    return node;
  }

  return getScrollableParent(node.parentNode) || null;
};

export const getInlinePaddingStart = (level: number, isSearchable: boolean) => {
  const initialPadding = isSearchable ? 0.25 : 1;

  return initialPadding + level * 0.75;
};

export const buildDropdownMenu = (list: DropdownItem[]): DropdownItem[] => {
  const hasNestedItems = list.some((item) => item.items && item.items.length > 0);

  if (!hasNestedItems) return list;

  const filteredGroups = list.filter((groupItem) => groupItem.items && groupItem.items.length > 0);

  return filteredGroups.reduce<DropdownItem[]>((result, groupItem, groupIndex) => {
    const isLastGroup = groupIndex === filteredGroups.length - 1;
    const groupItems = groupItem.items || [];
    const newGroup = [...groupItems];

    // Add divider to the last item of the group (except for the last group)
    if (!isLastGroup && newGroup.length > 0) {
      const lastItemIndex = newGroup.length - 1;
      newGroup[lastItemIndex] = { ...newGroup[lastItemIndex], divider: true };
    }
    return result.concat(newGroup);
  }, []);
};
