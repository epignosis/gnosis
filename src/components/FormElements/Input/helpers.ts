export const preventNonNumericalInput = (
  e: React.KeyboardEvent<HTMLInputElement>,
  isNumberType: boolean,
): void => {
  const allowedChars = ["Backspace", "Tab", ".", ",", "ArrowUp", "ArrowDown"];
  const numberPattern = /[0-9]/;

  const isCharacterNumber = numberPattern.test(e.key);
  const isAllowedChar = allowedChars.includes(e.key);

  if (!isCharacterNumber && !isAllowedChar && isNumberType) {
    e.preventDefault();
  }
};
