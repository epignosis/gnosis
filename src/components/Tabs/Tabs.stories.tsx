import React from "react";
import { Story } from "@storybook/react";
import { Heading } from "../../";
import Tabs from "./Tabs";

export default {
  title: "Components/Tabs",
};

const commonProps = {
  stickyHeader: false,
  responsiveHeader: false,
};

const Template: Story = (args) => (
  <Tabs {...args}>
    <Tabs.TabPane title="Completed courses">
      <div style={{ padding: "2rem" }}>
        <Heading>Completed courses</Heading>
        <p>Completed courses content!</p>
      </div>
    </Tabs.TabPane>
    <Tabs.TabPane title="My super points">
      <div style={{ padding: "2rem" }}>
        <Heading>My super points</Heading>
        <p>My super points content!</p>
      </div>
    </Tabs.TabPane>
    <Tabs.TabPane title="My superior level">
      <div style={{ padding: "2rem" }}>
        <Heading>My superior level</Heading>
        <p>My superior level content!</p>
      </div>
    </Tabs.TabPane>
    <Tabs.TabPane title="I have a lot of certifications">
      <div style={{ padding: "2rem" }}>
        <Heading>I have a lot of certifications</Heading>
        <p>I have a lot of certifications content!</p>
      </div>
    </Tabs.TabPane>
  </Tabs>
);

export const Basic = Template.bind({});

Basic.args = {
  ...commonProps,
};

export const ResponsiveHeader = Template.bind({});

ResponsiveHeader.args = {
  responsiveHeader: true,
};
