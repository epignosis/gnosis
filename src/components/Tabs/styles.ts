import { css, Theme, SerializedStyles } from "@emotion/react";
import { mq } from "@theme/utils/breakpoints";

export const container = css`
  position: relative;
  display: flex;
  flex-direction: column;
  flex: 1;

  #content {
    flex: 1;
    z-index: 0;
  }

  .nav-wrapper {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
`;

export const tabsHeader = (
  { tabs }: Theme,
  { stickyHeader }: { stickyHeader?: boolean },
): SerializedStyles => {
  return css`
    display: flex;
    flex-wrap: nowrap;
    padding: 0;
    white-space: nowrap;
    background: ${tabs.headerBackground};
    position: ${stickyHeader ? "sticky" : "static"};
    top: ${stickyHeader ? 0 : "auto"};
    z-index: ${stickyHeader ? 1 : 0};

    width: 100%;
    overflow-x: auto;

    ${mq["sm"]} {
      border-bottom: 1px solid ${tabs.headerBorder};
    }

    &::-webkit-scrollbar {
      display: none;
    }
  `;
};

export const tabNavItem = (
  { tabs }: Theme,
  { isActive }: { isActive: boolean },
): SerializedStyles => {
  return css`
    font-weight: ${isActive ? "700" : "400"};
    margin: 0 1rem 0 0;
    padding: 0.6rem;
    color: ${tabs.linkNormal};
    border-bottom: 4px solid ${isActive ? tabs.linkBorder : "transparent"};
  `;
};
