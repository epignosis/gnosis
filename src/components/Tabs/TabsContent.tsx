import React, { FC, ReactNode, CSSProperties } from "react";
import { tabsContent } from "./styles";

type TabsContentProps = {
  index: number;
  content: ReactNode;
  isVisible: boolean;
  style?: CSSProperties;
};

const TabsContent: FC<TabsContentProps> = ({ index, content, isVisible, style }) => {
  return (
    <article id={`content-${index}`} css={tabsContent(isVisible)} role="tabpanel" style={style}>
      {content}
    </article>
  );
};

export default TabsContent;
