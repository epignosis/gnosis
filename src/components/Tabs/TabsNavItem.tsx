import React, { FC, MouseEvent } from "react";
import { SerializedStyles } from "@emotion/react";
import classNames from "classnames";
import { tabNavItem } from "./styles";

type TabsNavItemProps = {
  index: number;
  title: string;
  isActive: boolean;
  onSelectTab: (i: number) => void;
};

const TabsNavItem: FC<TabsNavItemProps> = ({ index, title, isActive, onSelectTab }) => {
  const HtmlClasses = classNames({
    "tab-link": true,
    selected: isActive,
  });
  const onClick = (e: MouseEvent): void => {
    e.preventDefault();
    onSelectTab(index);
  };

  return (
    <a
      css={(theme): SerializedStyles => tabNavItem(theme, { isActive })}
      href={`#tab-${index}`}
      id={`tab-${index}`}
      aria-controls={`content-${index}`}
      role="tab"
      className={HtmlClasses}
      onClick={onClick}
    >
      {title}
    </a>
  );
};

export default TabsNavItem;
