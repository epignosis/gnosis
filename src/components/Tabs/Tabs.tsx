import React, { FC, useState, ReactElement, Children } from "react";
import { SerializedStyles } from "@emotion/react";
import { useResponsive } from "@umijs/hooks";
import TabsNavItem from "./TabsNavItem";
import TabsContent from "./TabsContent";
import { container, tabsHeader } from "./styles";
import { Select } from "@components";

type TabsProps = {
  stickyHeader?: boolean;
};

type TabPaneProps = {
  title: JSX.Element | string;
  fallbackTitle?: string;
};

const TabPane: FC<TabPaneProps> = () => null;

type TabsCompoundProps = {
  TabPane: FC<TabPaneProps>;
};

const Tabs: FC<TabsProps> & TabsCompoundProps = ({ children, stickyHeader }) => {
  const [activeTab, setActiveTab] = useState(0);
  const { sm } = useResponsive();
  const tabTitles = Children.map(children, (child, i) => ({
    index: i,
    title: (child as ReactElement).props.title,
    fallbackTitle:
      typeof (child as ReactElement).props.title === "string"
        ? (child as ReactElement).props.title
        : (child as ReactElement).props.fallbackTitle,
  }));
  const tabPanes = Children.map(children, (child, i) => ({
    index: i,
    content: (child as ReactElement).props.children,
  }));
  const onSelectTab = (index: number): void => {
    setActiveTab(index);
  };
  const isMobile = tabTitles?.length && sm;
  const isAboveMobile = tabTitles?.length && !sm;

  return (
    <section css={container}>
      <nav css={(theme): SerializedStyles => tabsHeader(theme, { stickyHeader })} role="tablist">
        {isMobile &&
          tabTitles?.map(({ index, title }) => (
            <TabsNavItem
              key={index}
              index={index}
              title={title}
              isActive={activeTab === index}
              onSelectTab={onSelectTab}
            />
          ))}

        {isAboveMobile && (
          <Select aria-label="select tab" onChange={(index): void => onSelectTab(parseInt(index))}>
            {tabTitles?.map(({ index, fallbackTitle }) => (
              <option key={index} value={index}>
                {fallbackTitle}
              </option>
            ))}
          </Select>
        )}
      </nav>

      <section id="content" aria-live="polite" role="region">
        {tabPanes?.length &&
          tabPanes.map(({ index, content }) => (
            <TabsContent
              key={index}
              index={index}
              content={content}
              isVisible={activeTab === index}
            />
          ))}
      </section>
    </section>
  );
};

Tabs.TabPane = TabPane;

export default Tabs;
