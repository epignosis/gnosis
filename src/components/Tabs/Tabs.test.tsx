import React from "react";
import { act } from "react-dom/test-utils";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Tabs from "./Tabs";
import { screen, render } from "@test-utils/render";
import { resizeWindow } from "@test-utils/helpers/windowResize";

const tab1Txt = faker.lorem.paragraph();
const tab2Txt = faker.lorem.paragraph();
const tab3Txt = faker.lorem.paragraph();

const tab1TitleTxt = faker.lorem.word();
const tab2TitleTxt = faker.lorem.word();
const tab3TitleTxt = faker.lorem.word();

const tab1FallbackTitle = faker.lorem.word();

describe("<Tabs/>", () => {
  it("renders correctly", () => {
    render(
      <Tabs>
        <Tabs.TabPane title={tab1TitleTxt}>{tab1Txt}</Tabs.TabPane>
        <Tabs.TabPane title={tab2TitleTxt}>{tab2Txt}</Tabs.TabPane>
        <Tabs.TabPane title={tab3TitleTxt}>{tab3Txt}</Tabs.TabPane>
      </Tabs>,
    );

    const titles = screen.getAllByRole("tab");
    const tab1Content = screen.getByText(tab1Txt);
    const tab1Title = screen.getByText(tab1TitleTxt);

    expect(titles).toHaveLength(3);
    expect(tab1Content).toHaveTextContent(tab1Txt);
    expect(tab1Title).toHaveTextContent(tab1TitleTxt);
  });

  it("changes tabs correctly", () => {
    render(
      <Tabs>
        <Tabs.TabPane title={tab1TitleTxt}>{tab1Txt}</Tabs.TabPane>
        <Tabs.TabPane title={tab2TitleTxt}>{tab2Txt}</Tabs.TabPane>
      </Tabs>,
    );

    const tab2 = screen.getByText(tab2TitleTxt);

    userEvent.click(tab2);

    const tab2Content = screen.getByText(tab2Txt);

    expect(tab2Content).toHaveTextContent(tab2Txt);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Tabs>
        <Tabs.TabPane title="Tab 1">Test tab 1</Tabs.TabPane>
        <Tabs.TabPane title="Tab 2">Test tab 2</Tabs.TabPane>
        <Tabs.TabPane title="Tab 3">Test tab 3</Tabs.TabPane>
      </Tabs>,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with stickyHeader=`true`", () => {
    const { container } = render(
      <Tabs stickyHeader>
        <Tabs.TabPane title="Tab 1">Test tab 1</Tabs.TabPane>
        <Tabs.TabPane title="Tab 2">Test tab 2</Tabs.TabPane>
        <Tabs.TabPane title="Tab 3">Test tab 3</Tabs.TabPane>
      </Tabs>,
    );

    expect(container).toMatchSnapshot();
  });
});

describe("<Tabs/> on mobile", () => {
  it("renders correctly", () => {
    render(
      <Tabs>
        <Tabs.TabPane title={<button>{tab1TitleTxt}</button>} fallbackTitle={tab1FallbackTitle}>
          {tab1Txt}
        </Tabs.TabPane>
        <Tabs.TabPane title={tab2TitleTxt}>{tab2Txt}</Tabs.TabPane>
        <Tabs.TabPane title={tab3TitleTxt}>{tab3Txt}</Tabs.TabPane>
      </Tabs>,
    );

    act(() => {
      resizeWindow(320, 500);
    });

    const select = screen.getByLabelText("select tab");
    const options = screen.getAllByRole("option");
    const fallbackTitle = screen.getByText(tab1FallbackTitle);

    expect(options).toHaveLength(3);
    expect(fallbackTitle).toHaveTextContent(tab1FallbackTitle);
    expect(select).toHaveValue("0");

    userEvent.selectOptions(select, "1");

    expect(select).toHaveValue("1");
  });
});
