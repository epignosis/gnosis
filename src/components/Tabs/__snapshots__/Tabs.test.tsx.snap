// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs/> matches snapshot 1`] = `
<div>
  <section
    class="css-a5ycro-container"
  >
    <nav
      class="css-1cv5t7l-tabsHeader-tabsHeader"
      role="tablist"
    >
      <a
        aria-controls="content-0"
        class="css-1v2oram-tabNavItem-tabNavItem"
        href="#tab-0"
        id="tab-0"
        role="tab"
      >
        Tab 1
      </a>
      <a
        aria-controls="content-1"
        class="css-1n1jyfz-tabNavItem-tabNavItem"
        href="#tab-1"
        id="tab-1"
        role="tab"
      >
        Tab 2
      </a>
      <a
        aria-controls="content-2"
        class="css-1n1jyfz-tabNavItem-tabNavItem"
        href="#tab-2"
        id="tab-2"
        role="tab"
      >
        Tab 3
      </a>
    </nav>
    <section
      aria-live="polite"
      id="content"
      role="region"
    >
      <article
        class="css-16ie0sg-tabsContent-tabsContent"
        id="content-0"
        role="tabpanel"
      >
        Test tab 1
      </article>
      <article
        class="css-1c08a5r-tabsContent-tabsContent"
        id="content-1"
        role="tabpanel"
      >
        Test tab 2
      </article>
      <article
        class="css-1c08a5r-tabsContent-tabsContent"
        id="content-2"
        role="tabpanel"
      >
        Test tab 3
      </article>
    </section>
  </section>
</div>
`;

exports[`<Tabs/> matches snapshot with stickyHeader=\`true\` 1`] = `
<div>
  <section
    class="css-a5ycro-container"
  >
    <nav
      class="css-1b46b0m-tabsHeader-tabsHeader"
      role="tablist"
    >
      <a
        aria-controls="content-0"
        class="css-1v2oram-tabNavItem-tabNavItem"
        href="#tab-0"
        id="tab-0"
        role="tab"
      >
        Tab 1
      </a>
      <a
        aria-controls="content-1"
        class="css-1n1jyfz-tabNavItem-tabNavItem"
        href="#tab-1"
        id="tab-1"
        role="tab"
      >
        Tab 2
      </a>
      <a
        aria-controls="content-2"
        class="css-1n1jyfz-tabNavItem-tabNavItem"
        href="#tab-2"
        id="tab-2"
        role="tab"
      >
        Tab 3
      </a>
    </nav>
    <section
      aria-live="polite"
      id="content"
      role="region"
    >
      <article
        class="css-16ie0sg-tabsContent-tabsContent"
        id="content-0"
        role="tabpanel"
      >
        Test tab 1
      </article>
      <article
        class="css-1c08a5r-tabsContent-tabsContent"
        id="content-1"
        role="tabpanel"
      >
        Test tab 2
      </article>
      <article
        class="css-1c08a5r-tabsContent-tabsContent"
        id="content-2"
        role="tabpanel"
      >
        Test tab 3
      </article>
    </section>
  </section>
</div>
`;
